lexoid/core/utils.py

```diff
@@ -121,19 +122,214 @@ def download_file(url: str, temp_dir: str) -> str:
+        heading_pattern (str): The heading pattern to split on (e.g., '##' or 'underline')
+
+    Returns:
+        List[Dict]: List of dictionaries containing metadata and content
+    """
+    structured_content = []
+
+    if heading_pattern == "underline":
+        # Split by underline headings
+        pattern = r"^([^\n]+)\n-+$"
+        sections = re.split(pattern, markdown_content, flags=re.MULTILINE)
+        # Remove empty sections and strip whitespace
+        sections = [section.strip() for section in sections if section.strip()]
+
+        # Handle content before first heading if it exists
+        if sections and not re.match(r"^[^\n]+\n-+$", sections[0], re.MULTILINE):
+            structured_content.append(
+                {
+                    "metadata": {"title": url, "page": "Introduction"},
+                    "content": sections.pop(0),
+                }
+            )
+
+        # Process sections pairwise (heading, content)
+        for i in range(0, len(sections), 2):
+            if i + 1 < len(sections):
+                structured_content.append(
+                    {
+                        "metadata": {"title": url, "page": sections[i]},
+                        "content": sections[i + 1],
+                    }
+                )
+    else:
+        # Split by hash headings
+        regex = f"^{heading_pattern} .*$"
+        sections = re.split(regex, markdown_content, flags=re.MULTILINE)
+        headings = re.findall(regex, markdown_content, flags=re.MULTILINE)
+
+        # Remove empty sections and strip whitespace
+        sections = [section.strip() for section in sections if section.strip()]
+
+        # Handle content before first heading if it exists
+        if len(sections) > len(headings):
+            structured_content.append(
+                {
+                    "metadata": {"title": url, "page": "Introduction"},
+                    "content": sections.pop(0),
+                }
+            )
+
+        # Process remaining sections
+        for heading, content in zip(headings, sections):
+            clean_heading = heading.replace(heading_pattern, "").strip()
+            structured_content.append(
+                {"metadata": {"title": url, "page": clean_heading}, "content": content}
+            )
+
+    return structured_content
+
+
+def read_html_content(url: str, raw: bool = False) -> Union[str, List[Dict]]:
+    """
+    Reads the content of an HTML page from the given URL and converts it to markdown or structured content.
 
     Args:
         url (str): The URL of the HTML page.
+        raw (bool): Whether to return raw markdown text or structured data.
 
     Returns:
-        str: The markdown content of the HTML page.
+        Union[str, List[Dict]]: Either raw markdown content or structured data with metadata and content sections.
     """
     response = requests.get(url)
     soup = BeautifulSoup(response.content, "html.parser")
```

Comment:
Sometimes decoding errors r thrown,
`Some characters could not be decoded, and were replaced with REPLACEMENT CHARACTER`

Maybe we need to set the encoding explicitly?
```
soup = BeautifulSoup(open(html_path, 'r'),"html.parser",from_encoding="iso-8859-1")
```

Reply:
Fixed.

Reply:
There were also a few bugs in the logic of handling `raw=True`. Should be fixed now.